src/commands/update.ts

```diff
@@ -0,0 +1,52 @@
+import { SlashCommandBuilder } from 'discord.js';
+import { exec as __unsafeExecuteCommand } from 'node:child_process';
+
+let numInvocations: number = 0;
+const info = new SlashCommandBuilder()
+	.setName('update')
+	.setDescription('Pulls the latest changes and restarts the bot')
+	.setDefaultMemberPermissions('0');
+
+export const update: GlobalCommand = {
+	info,
+	requiresGuild: false,
+	async execute({ replyPrivately, user, interaction }) {
+		const admin_ids = process.env['ADMINISTRATORS']?.split(',');
+		if (!admin_ids) {
+			// TODO: make this a UserMessageException
+			throw new Error('There is no ADMINISTRATORS variable. You must set ADMINISTRATORS in .env');
+		}
+
+		if (!admin_ids.includes(user.id)) {
+			// TODO: make this a UserMessageException
+			throw new Error(
+				'You do not have permission to perform this command. Contact the bot administrator.'
+			);
+		}
+
+		numInvocations += 1;
+		if (numInvocations > 1) {
+			throw new Error(
+				`Cannot run update, there are already ${numInvocations - 1} update invocations running`
+			);
+		}
+
+		await replyPrivately('Updating...');
+		await execAsync('npm run update');
+		await interaction.editReply('Finished updating. Restarting now.');
+		await execAsync('npm run restart');
+		numInvocations -= 1; // In theory, we should never arrive here. Including for testing purposes.
```

Comment:
Shouldn't we? I figure once the commands have been `await`-ed, if they didn't throw, we should arrive here.

At any rate, this assignment should probs go in the tail end of a `try...finally` block

Reply:
once `npm run restart` finished, this process will have been killed. I suppose in any case that my thought behind the comment doesn't really make sense. If we've somehow gotten here without the bot dying, we probably want to be able to run the command again. Good call on the try...finally, I'll add that.

Reply:
In that case, it might be prudent to wrap `execAsync('npm run restart')` in a function that returns `Promise<never>`, so that TypeScript is aware that we plan on dying before it resolves.

Reply:
With helpful comments explaining this weirdness, of course :P

Reply:
Perhaps something like this:

```ts
/**
 * Restarts the bot.
 */
async function restart(): Promise<never> {
	await execAsync('npm run restart');
	process.exit(); // so that TypeScript is *absolutely sure* we're dead after this
}
```

Reply:
I can wrap it in a function that returns `Promise<never>`, but typescript doesn't like that (according to it), the function has a reachable end point. What would you recommend doing to have it not be angry? I can have it throw after, which feels a little hacky, or maybe use an assertion comment.

Reply:
TIL that TypeScript doesn't realize that `await`-ing a `Promise<never>` results in unreachable code following that call. Maybe it will one day, idk. Seems a good idea nonetheless to have a `Promise` in there that semantically "`never` resolves."

Reply:
That `process.exit()` should never get hit, but if it does, at least we know the bot is dead dead by the end, lol

Reply:
Oh crap, the tests don't know about this. Would they even run to completion now?

Reply:
ah, good catch. No they don't

Reply:
I'm wondering now if we shouldn't just throw something that's like "We shouldn't have gotten here, lol"

Reply:
ok, I think I've got something that satisfies both the tests and production, and makes some sense to a reader. Throwing an error makes the tests overly complicated.